### ISSUE TITLE
[Router lite] attached Lifecycle hooks execution hierarchy

### ISSUE THREAD

**Issue by ivanbacher** [aurelia/aurelia#2019]
This might be a bug or expected behaviour. Not sure, so feel free to close if it's expected.

Example app:

```
MyApp
    RouteOne
        ComponentOne
            ComponentTwo
```

From reading the docs the order of the `attached` hook should be bottom up. So

```
ComponentTwo - attached()
ComponentOne  - attached()
RouteOne - attached()
MyApp  - attached()
```

However it is currently:

```
MyApp  - attached()
ComponentTwo - attached()
ComponentOne  - attached()
RouteOne - attached()
```

`MyApp  - attached()` gets called first. Maybe because it's the root component?

[StackBlitz example](https://stackblitz.com/edit/au2-attached-hook-hierarchy)

**Comment by Sayan751** [comment]
> MyApp  - attached() gets called first. Maybe because it's the root component?

@ivanbacher You are right. This is not a bug. A routed view always gets activated when the hosting `au-viewport` is attached. This explains the order of invocations, as you have observed.

I think the docs for router-lite can be adjusted to reflect this. In case you feel like contributing to the docs, feel free to submit a PR. Otherwise, I will try to push the changes in docs.

**Comment by bigopon** [comment]
Is this intuitive? I thought the viewport would only call attached after it has finished loading. Just asking based on the perspective of this issue.

**Comment by Sayan751** [comment]
@bigopon @ivanbacher I was not very clear with my explanation before. Let me add additional information.

The initial load of routed route happens post-activation of the app.

https://github.com/aurelia/aurelia/blob/0f3dbee6b350ef6b7288f901ac9d3df525125c0e/packages/router-lite/src/configuration.ts#L72

This means that even if the `au-viewport` in the root component is attached and the app gets completely activated first. At this point, the viewport is still without any routed view-model, as the initial load by the router is yet to happen. Once the app is activated, as a post-activation task, the router loads a route for the first time. And this is the reason behind the observed order of lifecycle hook invocations.

I am inclined towards keeping the order of invocations in its current state, unless it is causing a problem that cannot be solved without changing the order of invocations.

**Comment by bigopon** [comment]
An issue with that timing is we cannot tell when the initial routing has finished in app root `attached`, and likely will have to opt to listen to some events with a more cumbersome setup. It'd be better if we can have the initial timing more "smooth".
Or did I mis-analyze it?

**Comment by Sayan751** [comment]
We cannot start loading a route before the viewports are registered. You may say that we can observe when a viewport is registered and then start loading a route. However, that will be a hacky way to do this and might produce fragile code. Moreover, we don't know how many viewports are there in app-root; I mean technically, it is not a job for the router to analyse the app-root markup. For these reasons, I think making the initial timing "smoother" has diminishing return.

I would again like to ask if there is any particular problem that can only be solved by having a "smoother" timing? If there is none, I would like to update the docs to reflect this scenario.

**Comment by ivanbacher** [comment]
@Sayan751 Thanks for explaining. Makes total sense.

@bigopon 
> An issue with that timing is we cannot tell when the initial routing has finished in app root `attached`, and likely will have to opt to listen to some events with a more cumbersome setup. It'd be better if we can have the initial timing more "smooth". Or did I mis-analyze it?

Yes, this was how I noticed the behaviour. 
In short, I wanted to show a loading spinner in my app until all routes/child components were added. I assumed that if I added some code in the `attached` hook of my `MyApp` component that that would be called once everything was attached to the viewport.

Anyway, It's easy enough to use the EventAggregator or something similar overcome this. I think just added some notes to the docs that describe this behaviour would be helpful.
